packages/cli/src/commands/hydrogen/upgrade.ts

```diff
@@ -0,0 +1,2931 @@
+    return null;
+  }
+
+  const latestVersion = choices[0].value;
+
+  const selectedVersion = await renderSelectPrompt({
+    message: `Available Hydrogen versions (current: ${currentVersion})`,
+    choices: choices,
+    defaultValue: latestVersion,
+  });
+
+  const selectedRelease = upgradesAvailable.find(
+    (release) => release.version === selectedVersion,
+  );
+
+  if (!selectedRelease) {
+    throw new Error('No version selected');
+  }
+
+  return selectedRelease;
+}
+
+/**
+ * Displays a summary of the upgrade and next steps
+ */
+async function displaySummary({
+  currentVersion,
+  dryRun,
+  selectedRelease,
+  instrunctionsFilePath,
+}: {
+  currentVersion: string;
+  dryRun: boolean;
+  selectedRelease: Release;
+  instrunctionsFilePath?: string;
+}) {
+  const updatedDependenciesList = [
+    ...Object.entries(selectedRelease.dependencies || {}).map(
+      ([name, version]) => `${name}@${version}`,
+    ),
+    ...Object.entries(selectedRelease.devDependencies || {}).map(
+      ([name, version]) => `${name}@${version}`,
+    ),
+  ];
+
+  let nextSteps = [];
+
+  if (typeof instrunctionsFilePath === 'string') {
+    let instructions = `Upgrade instructions created at:\nfile://${instrunctionsFilePath}`;
+    nextSteps.push(instructions);
+  }
+
+  const releaseNotesUrl = `https://hydrogen.shopify.dev/releases/${selectedRelease.version}`;
+
+  nextSteps.push(`Release notes:\n${releaseNotesUrl}`);
+
+  const versionsChange = `${getAbsoluteVersion(
+    currentVersion,
+  )} → ${getAbsoluteVersion(`${selectedRelease.version}`)}`;
+
+  const headline = dryRun
+    ? `Dry run upgrade ${versionsChange}`
+    : `You've updated from ${versionsChange}`;
+
+  return renderSuccess({
+    headline,
+    customSections: [
+      {
+        title: 'Updated dependencies',
+        body: [
+          {
+            list: {
+              items: updatedDependenciesList,
+            },
+          },
+        ],
+      },
+      {
+        title: 'What’s next?',
+        body: [
+          {
+            list: {
+              items: nextSteps,
+            },
+          },
+        ],
+      },
+    ].filter(Boolean),
+  });
+}
+
+/**
+ * Validate if a h2 upgrade was successful by comparing the previous and current
+ * @shopify/hydrogen versions
+ */
+async function validateUpgrade({
+  appPath,
+  selectedRelease,
+}: {
+  appPath: string;
+  selectedRelease: Release;
+}) {
+  const dependencies = await getDependencies(
+    path.join(appPath, 'package.json'),
+  );
+
+  const updatedVersion = dependencies['@shopify/hydrogen'];
+
+  if (!updatedVersion) {
+    throw new Error('Hydrogen version not found in package.json');
+  }
+
+  const updatedPinnedVersion = getAbsoluteVersion(updatedVersion);
+
+  if (updatedPinnedVersion !== selectedRelease.version) {
+    renderError({
+      headline: `Failed to upgrade to Hydrogen version ${selectedRelease.version}`,
+      body: `You are still on version ${updatedPinnedVersion}`,
```

Comment:
Curious: is this code path triggered by a known issue? Or just trying to handle unknown issues with the upgrade?
Meaning, if `npm install xyz` fails, this code path won't run because it exits before entering here. So how would this code run, when `npm install xyz` succeeds but somehow the `package.json` is not updated?